### PR TITLE
Support IPv6 addresses during fingerprinting

### PR DESCRIPTION
**Pull request by diptanu**
This enables the Nomad fingerprinter to pick up IPv6 addresses. Once this is merged I will add a note in the documentation that Nomad may pick either an IPv4 or an IPv6 address on a dual stack system if an interface has both ipv4 and ipv6 addresses.

### CODE REVIEW

File: client/fingerprint/network.go [hunk 1]
```diff
@@ -69,66 +69,80 @@ func (f *NetworkFingerprint) Fingerprint(cfg *config.Config, node *structs.Node)
...
+	nwResources, err := f.createNetworkResources(mbits, intf)
+	if err != nil {
+		return false, err
 	}
 
-	node.Resources.Networks = append(node.Resources.Networks, newNetwork)
+	// Add the network resources to the node
+	for _, nwResource := range nwResources {
+		node.Resources.Networks = append(node.Resources.Networks, nwResource)
+		f.logger.Printf("[DEBUG] fingerprint.network: Detected interface %v with IP: %v, CIDR: %v during fingerprinting",
```
Comment (dadgar) [inline client/fingerprint/network.go]: Remove during fingerprinting. The prefix is `fingerprint.network` :)

File: client/fingerprint/network.go [hunk 2]
```diff
@@ -69,66 +69,80 @@ func (f *NetworkFingerprint) Fingerprint(cfg *config.Config, node *structs.Node)
...
+			Device: intf.Name,
+			MBits:  throughput,
+		}
+
+		// Find the IP Addr and the CIDR from the Address
 		var ip net.IP
 		switch v := (addr).(type) {
 		case *net.IPNet:
 			ip = v.IP
+			newNetwork.IP = v.IP.String()
```
Comment (dadgar) [inline client/fingerprint/network.go]: Doesn't this get overrides right below?

File: client/fingerprint/network.go [hunk 3]
```diff
@@ -69,66 +69,80 @@ func (f *NetworkFingerprint) Fingerprint(cfg *config.Config, node *structs.Node)
...
+	// Create the network resources from the interface
+	nwResources, err := f.createNetworkResources(mbits, intf)
+	if err != nil {
+		return false, err
 	}
 
-	node.Resources.Networks = append(node.Resources.Networks, newNetwork)
+	// Add the network resources to the node
+	for _, nwResource := range nwResources {
+		node.Resources.Networks = append(node.Resources.Networks, nwResource)
```
Comment (dadgar) [inline client/fingerprint/network.go]: `node.Resources.Networks =  nwResource` outside of the loop?

File: client/fingerprint/network.go [hunk 4]
```diff
@@ -69,66 +69,80 @@ func (f *NetworkFingerprint) Fingerprint(cfg *config.Config, node *structs.Node)
...
 		switch v := (addr).(type) {
 		case *net.IPNet:
 			ip = v.IP
+			newNetwork.IP = v.IP.String()
 		case *net.IPAddr:
 			ip = v.IP
 		}
+
+		// If the ip is link-local then we ignore it
+		if ip.IsLinkLocalUnicast() || ip.IsLinkLocalMulticast() {
```
Comment (dadgar) [inline client/fingerprint/network.go]: Can we get a test around these and IPv4/6 CIDRs

File: client/fingerprint/network.go [hunk 5]
```diff
@@ -69,66 +69,80 @@ func (f *NetworkFingerprint) Fingerprint(cfg *config.Config, node *structs.Node)
 		return false, nil
 	}
 
-	if ip, err = f.ipAddress(intf); err != nil {
-		return false, fmt.Errorf("Unable to find IP address of interface: %s, err: %v", intf.Name, err)
-	}
-
-	newNetwork.Device = intf.Name
```
Comment (dadgar) [inline client/fingerprint/network.go]: Need to set device name

Comment (diptanu) [inline client/fingerprint/network.go]: @dadgar Look at L118 where we are initializing the `structs.NetworkResource`

File: client/fingerprint/network.go [hunk 6]
```diff
@@ -95,6 +95,7 @@ func (f *NetworkFingerprint) Fingerprint(cfg *config.Config, node *structs.Node)
 		f.logger.Printf("[DEBUG] fingerprint.network: Detected interface %v with IP: %v", intf.Name, nwResource.IP)
 	}
 
+	// Deprectaed, setting the first IP as unique IP for the node
```
Comment (dadgar) [inline client/fingerprint/network.go]: Deprectaed -> Deprecated

**Pull request merged**